plugin/trino-hive/src/main/java/io/trino/plugin/hive/CachingDirectoryLister.java

```diff
@@ -279,4 +283,6 @@ private static boolean isLocationPresent(Storage storage)
         // Some Hive table types (e.g.: views) do not have a storage location
         return storage.getOptionalLocation().isPresent() && isNotEmpty(storage.getLocation());
     }
+
+    private static class ValuePlaceholder {}
```

Comment:
I was thinking about this a bit, and thought it could be nice to avoid the casts at all.

Let's model this:
```
ValueHolder {
  Optional<List<..>>  value;
}
```

and remove polymorphism from the code here.

also, document why we cannot use `Optional<List<..>>` directly in the cache.
-- we need object identity semantics for the value holder, not value-based class semantics